app/models/user.rb

```diff
@@ -11,17 +11,16 @@ class User < ApplicationRecord
   has_many :likes, dependent: :destroy
 
   has_many :friendships
-  has_many :inverse_friendships, class_name: 'Friendship', foreign_key: 'friend_id'
+  has_many :inverse_friendships, -> { where confirmed: nil }, class_name: 'Friendship', foreign_key: 'friend_id'
+  has_many :received_requests, through: :inverse_friendships
 
-  def friend?(user)
-    friends.include?(user)
-  end
+  has_many :accepted_friendships, -> { where confirmed: true }, class_name: 'Friendship'
+  has_many :friends, through: :accepted_friendships
 
-  private
+  has_many :sent_friendship_requests, -> { where confirmed: nil }, class_name: 'Friendship', foreign_key: 'user_id'
+  has_many :pending_friends, through: :sent_friendship_requests, source: :friend
 
-  def friends
-    friends_array = friendships.map { |friendship| friendship.friend if friendship.confirmed }
-    friends_array += inverse_friendships.map { |friendship| friendship.user if friendship.confirmed }
-    friends_array.compact
+  def friend?(user)
+    friends.include?(user)
   end
```

Comment:
* [ ] in order to get all posts written by user on their friends, you can use the simple method in User model (in order to NOT do it with Enumerables)

```ruby
# User model

def friends_and_own_posts
  Post.where(user: (self.friends.to_a << self))
  # This will produce SQL query with IN. Something like: select * from posts where user_id IN (1,45,874,43);
end
```

Reply:
Quite right.
This is a nice neat method for avoiding use of enumerables.
I have implemented it and the code is much cleaner ✔️